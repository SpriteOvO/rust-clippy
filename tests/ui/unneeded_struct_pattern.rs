//@aux-build:non-exhaustive-enum.rs
#![allow(clippy::manual_unwrap_or_default, clippy::manual_unwrap_or)]
#![warn(clippy::unneeded_struct_pattern)]

extern crate non_exhaustive_enum;
use non_exhaustive_enum::*;

fn main() {
    match Some(114514) {
        Some(v) => v,
        None {} => 0,
    };

    match Some(1919810) {
        Some(v) => v,
        None { .. } => 0,
    };

    match Some(123456) {
        Some(v) => v,
        None => 0,
    };

    match Some(Some(123456)) {
        Some(Some(v)) => v,
        Some(None {}) => 0,
        None {} => 0,
    };

    enum Custom {
        HasFields {
            field: i32,
        },
        HasBracketsNoFields {},
        NoBrackets,
        #[non_exhaustive]
        NoBracketsNonExhaustive,
        Init,
    };

    match Custom::Init {
        Custom::HasFields { field: value } => value, // Should be ignored
        Custom::HasBracketsNoFields {} => 0,         // Should be ignored
        Custom::NoBrackets {} => 0,                  // Should be fixed
        Custom::NoBracketsNonExhaustive {} => 0,     // Should be fixed
        _ => 0,
    };

    match Custom::Init {
        Custom::HasFields { field: value } => value, // Should be ignored
        Custom::HasBracketsNoFields { .. } => 0,     // Should be ignored
        Custom::NoBrackets { .. } => 0,              // Should be fixed
        Custom::NoBracketsNonExhaustive { .. } => 0, // Should be fixed
        _ => 0,
    };

    match Custom::Init {
        Custom::NoBrackets {} if true => 0, // Should be fixed
        _ => 0,
    };

    match Custom::Init {
        Custom::NoBrackets {} | Custom::NoBracketsNonExhaustive {} => 0, // Should be fixed
        _ => 0,
    };
}

fn external_crate() {
    use ExtNonExhaustiveVariant::*;

    match ExhaustiveUnit {
        // Expected
        ExhaustiveUnit => 0,
        _ => 0,
    };

    match ExhaustiveUnit {
        // Exhaustive variant, should be fixed
        ExhaustiveUnit { .. } => 0,
        _ => 0,
    };

    match ExhaustiveUnit {
        // Exhaustive variant, should be fixed
        ExhaustiveUnit {} => 0,
        _ => 0,
    };

    match ExhaustiveUnit {
        ExhaustiveUnit => 0,
        // vvvvv Non-exhaustive variants, should all be ignored
        Unit { .. } => 0,
        Tuple { 0: field, .. } => field,
        StructNoField { .. } => 0,
        Struct { field, .. } => field,
        _ => 0,
    };
}
