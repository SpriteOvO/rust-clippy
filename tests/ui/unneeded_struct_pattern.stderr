error: struct pattern is not needed for a unit variant
  --> tests/ui/unneeded_struct_pattern.rs:11:13
   |
LL |         None {} => 0,
   |             ^^^ help: remove the struct pattern
   |
   = note: `-D clippy::unneeded-struct-pattern` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(clippy::unneeded_struct_pattern)]`

error: struct pattern is not needed for a unit variant
  --> tests/ui/unneeded_struct_pattern.rs:16:13
   |
LL |         None { .. } => 0,
   |             ^^^^^^^ help: remove the struct pattern

error: struct pattern is not needed for a unit variant
  --> tests/ui/unneeded_struct_pattern.rs:26:18
   |
LL |         Some(None {}) => 0,
   |                  ^^^ help: remove the struct pattern

error: struct pattern is not needed for a unit variant
  --> tests/ui/unneeded_struct_pattern.rs:27:13
   |
LL |         None {} => 0,
   |             ^^^ help: remove the struct pattern

error: struct pattern is not needed for a unit variant
  --> tests/ui/unneeded_struct_pattern.rs:44:27
   |
LL |         Custom::NoBrackets {} => 0,                  // Should be fixed
   |                           ^^^ help: remove the struct pattern

error: struct pattern is not needed for a unit variant
  --> tests/ui/unneeded_struct_pattern.rs:45:40
   |
LL |         Custom::NoBracketsNonExhaustive {} => 0,     // Should be fixed
   |                                        ^^^ help: remove the struct pattern

error: struct pattern is not needed for a unit variant
  --> tests/ui/unneeded_struct_pattern.rs:52:27
   |
LL |         Custom::NoBrackets { .. } => 0,              // Should be fixed
   |                           ^^^^^^^ help: remove the struct pattern

error: struct pattern is not needed for a unit variant
  --> tests/ui/unneeded_struct_pattern.rs:53:40
   |
LL |         Custom::NoBracketsNonExhaustive { .. } => 0, // Should be fixed
   |                                        ^^^^^^^ help: remove the struct pattern

error: struct pattern is not needed for a unit variant
  --> tests/ui/unneeded_struct_pattern.rs:58:27
   |
LL |         Custom::NoBrackets {} if true => 0, // Should be fixed
   |                           ^^^ help: remove the struct pattern

error: struct pattern is not needed for a unit variant
  --> tests/ui/unneeded_struct_pattern.rs:63:27
   |
LL |         Custom::NoBrackets {} | Custom::NoBracketsNonExhaustive {} => 0, // Should be fixed
   |                           ^^^ help: remove the struct pattern

error: struct pattern is not needed for a unit variant
  --> tests/ui/unneeded_struct_pattern.rs:63:64
   |
LL |         Custom::NoBrackets {} | Custom::NoBracketsNonExhaustive {} => 0, // Should be fixed
   |                                                                ^^^ help: remove the struct pattern

error: struct pattern is not needed for a unit variant
  --> tests/ui/unneeded_struct_pattern.rs:79:23
   |
LL |         ExhaustiveUnit { .. } => 0,
   |                       ^^^^^^^ help: remove the struct pattern

error: struct pattern is not needed for a unit variant
  --> tests/ui/unneeded_struct_pattern.rs:85:23
   |
LL |         ExhaustiveUnit {} => 0,
   |                       ^^^ help: remove the struct pattern

error: aborting due to 13 previous errors

